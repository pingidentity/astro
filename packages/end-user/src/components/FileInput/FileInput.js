import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Button from '../Button';


/**
* @callback FileInput~onChange
*
* @param {object} fileInput
*    A reference to the file input
*/

/**
* @callback FileInput~renderSelectedContent
*
* @param {object} e
*    The ReactJS synthetic event object.
*/

/**
 * @class FileInput
 * @desc A file selection component
 *
 * @param {Array.<String>} [accept]
 *      An array of strings containing the accepted mime types or extensions.
 * @param {string} [className]
 *      CSS class(es) applied to the top-level element.
 * @param {string} [data-id='file-input']
 *      The data-id attribute value applied to the top-level element.
 * @param {FileInput~onChange} onChange
 *      The callback triggered when the file input changes value.
 * @param {FileInput~onRemove} onRemove
 *      The callback triggered when the remove button is clicked.
 * @param {FileInput~renderContentSelector} [renderContentSelector]
 *      A function that renders the content shown before a file is selected.
 * @param {FileInput~renderSelectedContent} [renderSelectedContent]
 *      A function that renders the content shown after a file is selected.
 * @param {string} [value]
 *      The name of the selected file
 */


export default class FileInput extends Component {

    static propTypes = {
        accept: PropTypes.arrayOf(PropTypes.string),
        'data-id': PropTypes.string,
        onChange: PropTypes.func,
        onRemove: PropTypes.func,
        renderContentSelector: PropTypes.func,
        renderSelectedContent: PropTypes.func,
        value: PropTypes.string,
    };

    static defaultProps = {
        accept: [],
        'data-id': 'file-input',
        onChange: () => {},
    };

    state = {
        hovered: false,
    };

    componentDidMount() {
        ['dragenter', 'dragleave', 'dragover', 'drop'].forEach((eventName) => {
            this.fileComponent.addEventListener(eventName, this._preventDefaults, false);
        });

        ['dragleave', 'drop'].forEach((eventName) => {
            this.fileComponent.addEventListener(eventName, this._onExit, false);
        });

        ['dragenter', 'dragover'].forEach((eventName) => {
            this.fileComponent.addEventListener(eventName, this._onHover, false);
        });

        ['drop'].forEach((eventName) => {
            this.fileComponent.addEventListener(eventName, this._handleDrop, false);
        });
    }

    /* istanbul ignore next  */
    componentWillUnmount() {
        ['dragenter', 'dragleave', 'dragover', 'drop'].forEach((eventName) => {
            this.fileComponent.removeEventListener(eventName, this._preventDefaults);
        });

        ['dragleave', 'drop'].forEach((eventName) => {
            this.fileComponent.removeEventListener(eventName, this._onExit);
        });

        ['dragenter', 'dragover'].forEach((eventName) => {
            this.fileComponent.removeEventListener(eventName, this._onHover);
        });

        ['drop'].forEach((eventName) => {
            this.fileComponent.removeEventListener(eventName, this._handleDrop);
        });
    }

    _handleDrop = (e) => {
        const {
            dataTransfer: {
                files: [file = null] = [],
            } = [],
        } = e;

        this._preventDefaults(e);
        this._handleFileChange(file, e);
    }

    _handleFileChange = (file, e) => {
        // needed for when canceling out of the file dialog
        if (!file) {
            return;
        }

        this.props.onChange(file, e);
    }

    _handleInputChange = (e) => {
        const file = e.target.files ? e.target.files[0] : null;
        this._handleFileChange(file, e);
    }

    _handleRemove = () => {
        // check for existance of ref if page is loaded with file pre-selected
        // (value is provided - ref not set yet)
        if (this.fileInput) {
            this.fileInput.value = null;
        }
        this.props.onRemove();
    }

    _onExit = () => this.setState({ hovered: false })

    _onHover = () => this.setState({ hovered: true })

    _preventDefaults = (e) => {
        e.preventDefault();
        e.stopPropagation();
    }

    _renderDefaultContentSelector = () => (
        <Button
            className="file-input--no-click"
            data-id="select-button"
            inline
            key="button"
            label="Select"
        />
    )

    _renderDefaultSelectedContent = () => (
        <div className="file-input__value">
            <span className="file-input__value-text" data-id="value">{this.props.value}</span>
            &nbsp;
            <Button
                className="file-input__remove-button"
                data-id="remove-button"
                inline
                label="Remove"
                onClick={this._handleRemove}
            />
        </div>
    )

    _renderContentSelector = () => (
        typeof this.props.renderContentSelector === 'function'
            ? this.props.renderContentSelector(this.props)
            : this._renderDefaultContentSelector(this.props)
    );

    _renderSelectedContent = () => (
        typeof this.props.renderSelectedContent === 'function'
            ? this.props.renderSelectedContent(this.props)
            : this._renderDefaultSelectedContent(this.props)
    );

    /* istanbul ignore next  */
    resetInput = () => {
        this.fileInput.value = '';
    }

    render() {
        const {
            accept,
            className,
            'data-id': dataId,
            value,
        } = this.props;
        const classNames = classnames('file-input', className, {
            'file-input--hovered': this.state.hovered,
        });

        return (
            <label
                className={classNames}
                data-id={dataId}
                ref={(div) => { this.fileComponent = div; }}
            >
                <input
                    accept={accept.join(',')}
                    data-id="input"
                    key="input"
                    name="input-file"
                    onChange={this._handleInputChange}
                    ref={(input) => { this.fileInput = input; }}
                    type="file"
                />
                {!value && this._renderContentSelector()}
                {value && this._renderSelectedContent()}
            </label>
        );
    }
}
