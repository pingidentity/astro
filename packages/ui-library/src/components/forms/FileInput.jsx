
import React, { Component } from "react";
import PropTypes from "prop-types";
import Button from "../buttons/Button";
import FileDrop from "./FileDrop";
import MessageButton from "../buttons/MessageButton";
import Icon from "../general/Icon";
import classnames from "classnames";

/**
* @callback FileInput~onRemove
*/

/**
* @callback FileDrop~onValueChange
*
* @param {Object} file
*     The file object
* @param {Object} e
*     The even object provided with the onChange event
*/

/**
* @callback FileInput~onValidateFile
*
* @param {boolean} isValidFile
*     A boolean indicating whether the file type is one the accepted file types.
* @param {Object} file
*     The file object
* @param {Object} e
*     The even object provided with the onChange event
*/

/**
 * @class FileInput
 * @desc A single-selection component with big icons and titles.
 *
 * @param {string} [data-id="file-drop"]
 *     Defines the "data-id" for top-level HTML container.
 * @param {string[]} [accept]
 *     An optional array of the the allowed file mime types or file extensions.]
 * @param {boolean} [loading]
 *     If the component should show a loading state for uploading files
 * @param {boolean} [noBorder]
 *     If the component shows a line around the upload this will remove it.
 * @param {string} [className]
 *     Optional CSS classname(s) applied to top-level container.
 * @param {string} [fileName]
 *     The name of the currently selected file.
 * @param {FileInput~onRemove} [onRemove]
 *     The callback triggered when the remove file link is clicked.
 * @param {FileInput~onValueChange} [onValueChange]
 *     A callback triggered when a file is dropped or selected.
 * @param {FileInput~onValidateFile} [onValidateFile]
 *     The callback triggered when when a file is selected that idicates if the selected file type matches those allowed by the "accept" prop.
 * @param {Object} [strings]
 *     An object containing the various blurbs of text rendered in the component.
 *
 * @example
 *     <FileInput
 *         onValidate={this._handleValidate}
 *         onValueChange={this._handleChange}
 *         onRemove={this._handleRemove}
 *         fileName={this.state.selectedFile}
 *         accept={["text/csv", "image/jpeg", "image/png"]}
 *     />
 */

export default class FileInput extends Component {

    static propTypes = {
        "data-id": PropTypes.string,
        accept: PropTypes.array,
        className: PropTypes.string,
        fileName: PropTypes.string,
        fileData: PropTypes.node,
        onRemove: PropTypes.func,
        noBorder: PropTypes.bool,
        onValidateFile: PropTypes.func,
        onValueChange: PropTypes.func,
        strings: PropTypes.objectOf(PropTypes.string),
        status: PropTypes.shape({
            label: PropTypes.string,
            type: PropTypes.string,
        }),
    };

    static defaultProps = {
        strings: {},
        noBorder: false,
    }

    defaultStrings = {
        select: "Choose",
        remove: "Remove",
    };

    _clickButton = (inputRef) => /* istanbul ignore next  */ () => inputRef.click();

    _handleRemove = (inputRef) => () => {
        inputRef.value = null;
        this.props.onRemove();
    }

    _renderContent = (props) => {
        const {
            inputRef
        } = props;
        const {
            fileName,
            fileData,
            noBorder,
            selectedTitle,
            strings,
            status,
        } = this.props;
        const text = { ...this.defaultStrings, ...strings };

        const classNames = classnames(this.props.className, "input-file__field-set", {
            "unfocused": noBorder,
            "focused": !noBorder,
            "input-file__field-set--focused": !noBorder,
        });

        return (
            fileName && !status ? (
                <fieldset className={classNames}>
                    <legend>{selectedTitle}</legend>
                    <div className="input-file__selected-content">
                        <div className="input-file__info">
                            <Icon
                                data-id="file-icon"
                                className="input-file__file-icon"
                                iconName="docs"
                                iconSize={Icon.iconSizes.MD}
                                type="leading"
                            />
                            <div className="input-file__file">
                                <div data-id="file-name" className="input-file__file-name">
                                    {fileName}
                                </div>
                                {fileData && (
                                    <div data-id="file-data" className="input-file__file-data">{fileData}</div>
                                )}
                            </div>
                        </div>
                        <Button
                            data-id="remove-button"
                            className="input-file__remove-btn"
                            onClick={this._handleRemove(inputRef)}
                            inline
                        >
                            {text.remove}
                        </Button>
                    </div>
                </fieldset>
            ) : (
                <div className="input-file__select-btn">
                    <MessageButton
                        status={status ? status.type : "default" }
                        data-id="select-button"
                        onClick={this._clickButton(inputRef)}
                        label={status ? status.label : text.select}
                        inline
                    />
                </div>
            )

        );
    }

    render () {
        const {
            fileName,
        } = this.props;

        const classNames = classnames(this.props.className, "input-file", {
            "input-file--selected": fileName,
        });

        return (
            <FileDrop
                {...this.props}
                renderContent={this._renderContent}
                className={classNames}
            />
        );
    }
}
