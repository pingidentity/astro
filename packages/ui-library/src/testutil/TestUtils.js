/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
