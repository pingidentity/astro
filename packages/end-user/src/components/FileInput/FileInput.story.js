import React from 'react';
import FileInput from './FileInput';

export default {
    title: 'Components/Inputs/FileInput',
    component: FileInput,
};

export const Default = () => (
    <FileInput />
);
