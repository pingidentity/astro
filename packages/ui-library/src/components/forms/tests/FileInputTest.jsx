
import React from "react";
import ReactTestUtils from "react-dom/test-utils";
import TestUtils from "../../../testutil/TestUtils";
import FileInput from "../FileInput";
import { mount } from "enzyme";


describe("FileInput", function () {
    // const FileInput = require("../FileInput");

    const testDataId = "myinput";

    const customStrings = {
        select: "Select it",
        remove: "Remove it",
    };

    const defaultProps = {
        "data-id": testDataId,
        onRemove: jest.fn(),
        strings: customStrings,
    };

    function getComponent(opts) {
        return ReactTestUtils.renderIntoDocument(<FileInput {...defaultProps} {...opts} />);
    }

    it("renders the component in initial state", function () {
        const component = getComponent();
        const componentDom = TestUtils.findRenderedDOMNodeWithDataId(component, testDataId);
        const selectBtn = TestUtils.findRenderedDOMNodeWithDataId(component, "select-button");

        expect(componentDom).toBeTruthy();
        expect(componentDom.className).toContain("input-file");
        expect(selectBtn.textContent).toBe(customStrings.select);

    });

    it("renders custom in selected state", function () {
        const fileName = "test.jpg";
        const fileData = "some data";
        const component = getComponent({
            fileData: fileData,
            fileName: fileName,
        });
        const name = TestUtils.findRenderedDOMNodeWithDataId(component, "file-name");
        const data = TestUtils.findRenderedDOMNodeWithDataId(component, "file-data");
        const removeBtn = TestUtils.findRenderedDOMNodeWithDataId(component, "remove-button");

        expect(name.textContent).toContain(fileName);
        expect(data.textContent).toContain(fileData);
        expect(removeBtn.textContent).toContain(customStrings.remove);
    });

    it("renders the component with loading state", function () {
        const component = getComponent({
            loading: true
        });
        const loadingBtn = TestUtils.findRenderedDOMNodeWithDataId(component, "ellipsis-loader");

        expect(loadingBtn).toBeDefined();

    });

    it("calls the onRemove callback", function () {
        const myOnRemove = jest.fn();
        const wrapper = mount(
            <FileInput onRemove={myOnRemove} />
        );

        wrapper.setProps({ fileName: "foobar.png" });

        const removeBtn = wrapper.find(`button[data-id="remove-button"]`);

        expect(myOnRemove).not.toHaveBeenCalled();
        removeBtn.simulate("click");
        expect(myOnRemove).toHaveBeenCalled();

    });

    it("Renders the the fieldset without a border", () => {
        const component = mount(
            <FileInput fileName="hello" noBorder/>
        );
        const element = component.find(".input-file__field-set.unfocused");
        expect(element.exists()).toEqual(true);
    });

    it("Renders the fieldset with the title always shown", () => {
        const component = mount(
            <FileInput alwaysShowTitle />
        );
        const element = component.find(".input-file__always-title-container");
        expect(element.exists()).toEqual(true);
    });

    it("Renders the fieldset with the same title as the selected title", () => {
        const component = mount(
            <FileInput alwaysShowTitle selectedTitle="hello"/>
        );
        const element = component.find(".input-file__always-title");
        expect(element.exists()).toEqual(true);
    });
});
