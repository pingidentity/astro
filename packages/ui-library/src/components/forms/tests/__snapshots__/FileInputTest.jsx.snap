// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput data-id's don't change 1`] = `
Array [
  "input-filedrop",
  "select-button",
  "input-filedrop-label",
  "input-filedrop-input",
]
`;
